.github/workflows/delegator_generic_slsa3.yml

```diff
@@ -0,0 +1,257 @@
+        required: false
+      secret2:
+        description: >
+          Optional secret.
+
+          This argument is passed, unchanged, to the builder's Action's `build-artifacts`.
+        required: false
+
+    # Inputs provided as arguments are passed as inputs to the callback Action,
+    # formatted as a map.
+    inputs:
+      slsa-token:
+        description: "The SLSA token identifying the request"
+        required: true
+        type: string
+
+    # Outputs are provided to the caller TRW.
+    outputs:
+      # This contains all the ouputs of the callback Action, formatted as a map.
+      build-artifacts-outputs:
+        description: "The outputs from the build-artitacts Action, unchanged."
+        value: ${{ jobs.build-artifacts-ubuntu.outputs.outputs }}
+      
+      # This is an output from the framework.
+      attestations-download-name:
+        description: "Name of the artifact to download all the attestations."
+        value: ${{ jobs.generate-provenance.outputs.attestations-download-name }}
+
+jobs:
+  # rng generates a random number to avoid name collision in artifacts
+  # when multiple workflows run concurrently.
+  rng:
+    outputs:
+      value: ${{ steps.rng.outputs.random }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Generate random 16-byte value (32-char hex encoded)
+        id: rng
+        uses: slsa-framework/slsa-github-generator/.github/actions/rng@main
+
+  # verify-token verifies the slsa token.
+  verify-token:
+    outputs:
+      slsa-verified-token: ${{ steps.verify.outputs.slsa-verified-token }}
+      tool-repository: ${{ steps.verify.outputs.tool-repository }}
+      tool-ref: ${{ steps.verify.outputs.tool-ref }}
+      predicate-sha256: ${{ steps.upload.outputs.sha256 }}
+    runs-on: ubuntu-latest
+    needs: [rng]
+    steps:
+      - name: Verify token with test action
+        id: verify
+        uses: slsa-framework/slsa-github-generator/.github/actions/verify-token@main
+        with:
+          slsa-workflow-recipient: "delegator_generic_slsa3.yml"
+          slsa-unverified-token: ${{ inputs.slsa-token }}
+          output-predicate: ${{ env.SLSA_PREDICATE_FILE }}
+
+      - name: Debug output predicate
+        env: 
+          PREDICATE_FILE: ${{ env.SLSA_PREDICATE_FILE }}
+          TOKEN: ${{ toJson(steps.verify.outputs.slsa-verified-token) }}
+          RUNNER: ${{ fromJson(steps.verify.outputs.slsa-verified-token).builder.runner_label }}
+        run: |
+          cat "$PREDICATE_FILE" |  jq -r
+          echo "TOKEN: $TOKEN"
+          echo "RUNNER: $RUNNER"
+
+      - name: Upload predicate
+        id: upload
+        uses: slsa-framework/slsa-github-generator/.github/actions/secure-upload-artifact@main
+        with:
+          name: "${{ needs.rng.outputs.value }}-${{ env.SLSA_PREDICATE_FILE }}"
+          path: ${{ env.SLSA_PREDICATE_FILE }}
+
+  # privacy-check verifies that the user has agreed for their repository name to be made public, via the rekor log.
```

Comment:
do we not need detect-env to be performing that? or is that part of the TODO to update refs?

Reply:
If we refer to the Actions by ref, we actually don't need the `detect-env` Action. We just need to be careful to update the ref pre-submits.

Reply:
Hmm I think I see what you mean. if user calls `builder@tag`, then we have already hard-coded that reusable workflow at that ref to call actions at that tag in the builder repo.

And it was difficult I think to do this before because we would get stale pinned actions and so on during development.

Reply:
one problem before was also that we were calling scripts, and scripts cannot be referred by 'tag'. We could probably change the behavior in other workflows now